casper/src/main/scala/coop/rchain/casper/engine/Running.scala

```diff
@@ -241,37 +186,58 @@ object Running {
       _ <- maybeBlock match {
             case None => Log[F].info(logIntro + "No response given since block not found.")
             case Some(block) =>
-              streamToPeer(peer)(ToPacket(block.toProto)) >> Log[F].info(
+              CommUtil[F].streamToPeer(peer, ToPacket(block.toProto)) >> Log[F].info(
                 logIntro + "Response sent."
               )
           }
     } yield ()
 
-  def handleForkChoiceTipRequest[F[_]: Sync: RPConfAsk: Log: TransportLayer: BlockStore](
+  /**
+    * Peer asks if this node has particular block
+    */
+  def handleHasBlockRequest[F[_]: Monad: RPConfAsk: TransportLayer: CommUtil](
+      peer: PeerNode,
+      hbr: HasBlockRequest
+  )(blockLookup: BlockHash => F[Boolean]): F[Unit] =
+    blockLookup(hbr.hash).flatMap(
+      present =>
+        Applicative[F].whenA(present)(
+          CommUtil[F].sendToPeer(peer, HasBlockProto(hbr.hash))
+        )
+    )
+
+  /**
+    * Peer asks for fork-choice tip
+    */
+  def handleForkChoiceTipRequest[F[_]: Sync: RPConfAsk: Log: TransportLayer: BlockStore: CommUtil](
       peer: PeerNode,
       fctr: ForkChoiceTipRequest.type
   )(casper: MultiParentCasper[F]): F[Unit] =
-    Log[F].info(s"Received ForkChoiceTipRequest from $peer") >> MultiParentCasper.forkChoiceTip(
-      casper
-    ) >>= (
+    Log[F].info(s"Received ForkChoiceTipRequest from ${peer.endpoint.host}") >> MultiParentCasper
+      .forkChoiceTip(
+        casper
+      ) >>= (
         tip =>
           Log[F].info(
-            s"Streaming block ${PrettyPrinter.buildString(tip.blockHash)} to $peer"
-          ) >> streamToPeer(peer)(ToPacket(tip.toProto))
+            s"Streaming block ${PrettyPrinter.buildString(tip.blockHash)} to ${peer.endpoint.host}"
+          ) >> CommUtil[F].streamToPeer(peer, ToPacket(tip.toProto))
```

Comment:
I love combinators but if they are sprinkled everywhere in the code it's difficult to follow the execution. Just by assigning explicit names can make things much more readable and gives more insight what are the "dependencies" to execute the final expression.
```scala
val logRequestedTip = Log[F].info(s"Received ForkChoiceTipRequest from ${peer.endpoint.host}")
def logStreamingBlock(block: BlockMessage) = Log[F].info(
  s"Streaming block ${PrettyPrinter.buildString(block.blockHash)} to ${peer.endpoint.host}"
)
val forkChoiceTip = MultiParentCasper.forkChoiceTip(casper)
def streamToPears(tip: BlockMessage) =
  logStreamingBlock(tip) >> CommUtil[F].streamToPeer(peer, ToPacket(tip.toProto))

logRequestedTip >> forkChoiceTip >>= streamToPears
```
This is also why we need functions to be "referentially transparent" so that expression
```scala
val forkChoiceTip = MultiParentCasper.forkChoiceTip(casper)
```
will not actually do anything until used/run in monadic sequence.

Reply:
I've adopted this approach in Running object